File: Tyche.Android/ui/src/main/java/com/felipearpa/tyche/ui/state/LoadableViewState.kt
package com.felipearpa.tyche.ui.state

sealed class LoadableViewState<out Value : Any> {
    data object Initial : LoadableViewState<Nothing>()

    data object Loading : LoadableViewState<Nothing>()

    data class Success<Value : Any>(val value: Value) : LoadableViewState<Value>() {
        operator fun invoke(): Value = value
    }

    data class Failure(val exception: Throwable) : LoadableViewState<Nothing>() {
        operator fun invoke(): Throwable = exception
    }
}

fun <Value : Any> LoadableViewState<Value>.isInitial() = this is LoadableViewState.Initial

fun <Value : Any> LoadableViewState<Value>.isLoading() = this is LoadableViewState.Loading

fun <Value : Any> LoadableViewState<Value>.isSuccess() = this is LoadableViewState.Success

fun <Value : Any> LoadableViewState<Value>.isFailure() = this is LoadableViewState.Failure

inline fun <T : Any> LoadableViewState<T>.onInitial(block: () -> Unit): LoadableViewState<T> {
    if (isInitial()) {
        block()
    }
    return this
}

inline fun <Value : Any> LoadableViewState<Value>.onLoading(block: () -> Unit): LoadableViewState<Value> {
    if (isLoading()) {
        block()
    }
    return this
}

inline fun <Value : Any> LoadableViewState<Value>.onSuccess(block: (value: Value) -> Unit): LoadableViewState<Value> {
    if (isSuccess()) {
        block((this as LoadableViewState.Success).invoke())
    }
    return this
}

inline fun <Value : Any> LoadableViewState<Value>.onFailure(block: (throwable: Throwable) -> Unit): LoadableViewState<Value> {
    if (isFailure()) {
        block((this as LoadableViewState.Failure).invoke())
    }
    return this
}

fun <Value : Any> LoadableViewState<Value>.valueOrNull(): Value? {
    if (isSuccess())
        return (this as LoadableViewState.Success).invoke()
    return null
}

fun <Value : Any> LoadableViewState<Value>.exceptionOrNull(): Throwable? {
    if (isFailure())
        return (this as LoadableViewState.Failure).invoke()
    return null
}

File: Tyche.iOS/UI/Sources/UI/State/LoadableViewState.swift
public enum LoadableViewState<Value> {
    case initial
    case loading
    case success(Value)
    case failure(Error)
}

public extension LoadableViewState {
    @inlinable
    func isInitial() -> Bool {
        if case .initial = self {
            return true
        }
        return false
    }
    
    @inlinable
    func isLoading() -> Bool {
        if case .loading = self {
            return true
        }
        return false
    }
    
    @inlinable
    func isSuccess() -> Bool {
        if case .success = self {
            return true
        }
        return false
    }
    
    @inlinable
    func isFailure() -> Bool {
        if case .failure = self {
            return true
        }
        return false
    }
    
    @inlinable
    func valueOrNil() -> Value? {
        if case .success(let value) = self {
            return value
        }
        return nil
    }
    
    @inlinable
    func errorOrNil() -> Error? {
        if case .failure(let error) = self {
            return error
        }
        return nil
    }
}
